### PR TITLE
Remove string content from `rustls::Error`

### PR DESCRIPTION
**Pull request by ctz**
- [x] `PeerMisbehavedError` -> enum
- [x] `PeerIncompatibleError` -> enum
- [ ] `General` -> `Box<dyn Error>`
  - and audit uses of this
- [x] re-generalise certificate errors to  `Box<dyn Error>`
- [x] fix the README/lib.rs documentation that tells you to expect `Err(WebPkiError(CertExpired, ValidateServerCert))`

This is in pursuit of:

- #1162 
- #932 
- #930
- #889

### CODE REVIEW

**Comment by djc** [comment]
Nice! I would maybe drop the Reason suffix from the nested enum and impl From<PeerMisbehaved> for Error, so that you can call .into() and minimize the boilerplate.

**Comment by ctz** [comment]
Good idea! Have done that.

**Comment by djc** [comment]
If we're improving error stuff, might also drop the `Error` suffix from `Error` variant names.

**Comment by djc** [comment]
Also maybe do this as one PR per bullet point? Bit easier to digest and not much point in heaping them all together anyway.

**Comment by ctz** [comment]
> Also maybe do this as one PR per bullet point? Bit easier to digest and not much point in heaping them all together anyway.

can definitely do that when we come to merge this; at the moment the second commit depends on the first so that's going to have to happen in order

**Comment by djc** [comment]
> > Also maybe do this as one PR per bullet point? Bit easier to digest and not much point in heaping them all together anyway.
> 
> 
> 
> can definitely do that when we come to merge this; at the moment the second commit depends on the first so that's going to have to happen in order

Sure that's fine, so is this Ready for review?

**Comment by ctz** [comment]
> * [ ]  re-generalise certificate errors to  `Box<dyn Error>`

Unfortunately seems not to be possible because `std::error::Error` does not implement `Clone` or `PartialEq` which we require for it to be part of our error type.

So I think these are best dealt with as another sub-enum?

**Comment by djc** [comment]
We probably want to wrap a struct type around webpki error types, since just an enum variant can't hide the inner types.

**Comment by ctz** [comment]
> We probably want to wrap a struct type around webpki error types, since just an enum variant can't hide the inner types.

I think we probably don't want the webpki error type involved at all? Any involvement of it at all means someone bringing their own `ServerCertVerifier` requires it to be publicly constructible (so they can express errors with the certificate) which means webpki types leak out the crate anyway, and forces everyone into a webpki dependency. (Though I think DnsName already does that and is unconditional.) If the webpki type is involved, better to just stick it straight in the rustls Error enum.

At this point I'm minded to just have rustls define its own `CertificateError` enum and funnel everything through that.

> Or we could propagate the bounds? Presumably the nested error types may also implement Clone?

Ah, that would be good. Seems like doing:

```
pub trait CloneableError : StdError + Clone {}
```

gets:

```
error[E0038]: the trait `CloneableError` cannot be made into an object
  --> rustls/src/error.rs:78:28
   |
78 |     InvalidCertificate(Box<dyn CloneableError>),
   |                            ^^^^^^^^^^^^^^^^^^ `CloneableError` cannot be made into an object
   |
note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
```

Which is not too surprising, since asking a trait object to clone itself is tricky. Same goes for `PartialEq` (so even avoiding `Clone` through `Rc`/`Arc` won't work)

**Comment by djc** [comment]
> I think we probably don't want the webpki error type involved at all? Any involvement of it at all means someone bringing their own `ServerCertVerifier` requires it to be publicly constructible (so they can express errors with the certificate) which means webpki types leak out the crate anyway, and forces everyone into a webpki dependency. (Though I think DnsName already does that and is unconditional.) If the webpki type is involved, better to just stick it straight in the rustls Error enum.

I meant wrapping webpki errors into a `Box<dyn Error>`. But deriving a local `CertificateError` from that seems like a decent option, too.

> At this point I'm minded to just have rustls define its own `CertificateError` enum and funnel everything through that.

Yeah, probably makes sense.

Maybe we could just drop the `Clone` and `PartialEq` derives? I don't think that they're that commonly needed anyway.

**Comment by ctz** [comment]
> Maybe we could just drop the `Clone` and `PartialEq` derives? I don't think that they're that commonly needed anyway -- ah, I do think we have some internal usage of `clone()` in places where we want to fuse the connection.

Yeah I think we need clone. But I think we can get rid of `PartialEq` -- working on that now. `Clone` can be obviated by using `Arc` rather than `Box`.

File: rustls/src/error.rs [hunk 1]
```diff
@@ -60,17 +60,20 @@ pub enum Error {
...
+    /// We saw an invalid certificate.
+    ///
+    /// This includes, but is not limited to:
+    /// - we received an invalidly encoded certificate,
+    /// - we received a certificate that was not signed correctly
+    ///   by its alleged issuer,
+    /// - we received a certificate that was not issued by any
+    ///   of the configured trusted root certificates,
+    /// - we received a certificate that was not valid for the expect
+    ///   server name.
```
Comment (djc) [inline rustls/src/error.rs:72]: I probably wouldn't keep this list here, just refer to the docs for `CertificateError`?

File: rustls/src/verify.rs [hunk 2]
```diff
@@ -617,12 +617,22 @@ impl ClientCertVerifier for AllowAnyAnonymousOrAuthenticatedClient {
...
+        BadDer | BadDerTime => CertificateError::BadEncoding.into(),
+        CertNotValidYet => CertificateError::NotValidYet.into(),
+        CertExpired | InvalidCertValidity => CertificateError::Expired.into(),
+        UnknownIssuer => CertificateError::UnknownIssuer.into(),
+        CertNotValidForName => CertificateError::NotValidForName.into(),
+
+        InvalidSignatureForPublicKey
+        | UnsupportedSignatureAlgorithm
+        | UnsupportedSignatureAlgorithmForPublicKey => CertificateError::BadSignature.into(),
+        e => {
```
Comment (djc) [inline rustls/src/verify.rs:629]: Add a comment about `webpki::Error` being `non_exhaustive` and that this should thus be very unlikely? (Assuming you handled all the known variants above.)

File: rustls/src/error.rs (hunk 1 above)
Comment (ctz) [reply rustls/src/error.rs:72]: yeah will do that!

File: rustls/src/verify.rs (hunk 2 above)
Comment (ctz) [reply rustls/src/verify.rs]: i don't think this handles the known webpki errors, but just now i've pushed something that stops discarding the underlying error in the `Other` case

**Pull request merged**